packages/fxa-settings/src/pages/Signup/container.spec.tsx

```diff
@@ -0,0 +1,439 @@
+function mockIntegration() {
+  integration = {
+    type: IntegrationType.SyncDesktop,
+    getService: () => MozServices.Default,
+    features: {
+      allowUidChange: false,
+      fxaStatus: false,
+      handleSignedInNotification: false,
+      reuseExistingSession: false,
+      supportsPairing: false,
+      webChannelSupport: false,
+    },
+  };
+}
+let serviceName: MozServices;
+function mockServiceName() {
+  serviceName = MozServices.Default;
+}
+
+// TIP - Note that we don't actually want to render the Signup or LoadingSpinner react components,
+// since we aren't testing their logic. There fore it's fine to just return something super
+// simple. And check assumptions about props if necessary.
+let currentSignupProps: SignupProps | undefined;
+function mockSignupModule() {
+  currentSignupProps = undefined;
+  jest
+    .spyOn(SignupModule, 'Signup')
+    .mockImplementation((props: SignupProps) => {
+      currentSignupProps = props;
+      return <div>sign up mock</div>;
+    });
+}
+function mockLoadingSpinnerModule() {
+  jest.spyOn(LoadingSpinnerModule, 'LoadingSpinner').mockImplementation(() => {
+    return <div>loading spinner mock</div>;
+  });
+}
+
+// TIP - Most modules can be easily mocked via jest.spyOn. As you can see this very clean.
+function mockUseValidateModule() {
+  jest.spyOn(UseValidateModule, 'useValidatedQueryParams').mockReturnValue({
+    queryParamModel: {
+      email: 'foo@bar.com',
+    } as unknown as ModelDataProvider,
+    validationError: undefined,
+  });
+}
+
+function mockFirefoxModule() {
+  FirefoxModule.firefox.addEventListener = jest.fn();
+  FirefoxModule.firefox.send = jest.fn();
+}
+
+function mockCryptoModule() {
+  jest.spyOn(CryptoModule, 'getCredentials').mockResolvedValue({
+    authPW: 'apw123',
+    unwrapBKey: 'ubk123',
+  });
+}
+
+function mockReachRouterModule() {
+  jest.spyOn(ReachRouterModule, 'useNavigate').mockReturnValue(function () {
+    return Promise.resolve();
+  });
+}
+
+function mockReactUtilsModule() {
+  jest
+    .spyOn(ReactUtils, 'hardNavigateToContentServer')
+    .mockImplementation(() => {});
+}
+
+// TIP - Sometimes it's useful to have inner mocks. In this case, we hold a reference to
+// mockBeginSignupMutation, so it can be asserted against and redefined as needed.
+let mockBeginSignupMutation = jest.fn();
+function mockApolloClientModule() {
+  mockBeginSignupMutation.mockImplementation(async () => {
+    return {
+      data: {
+        unwrapBKey: 'foo',
+        SignUp: {
+          uid: 'uid123',
+          keyFetchToken: 'kft123',
+          sessionToken: 'st123',
+        },
+      },
+    };
+  });
+
+  jest.spyOn(ApolloModule, 'useMutation').mockReturnValue([
+    async (...args: any[]) => {
+      return mockBeginSignupMutation(...args);
+    },
+    {
+      loading: false,
+      called: true,
+      client: {} as ApolloClient<any>,
+    },
+  ]);
+}
+
+// TIP - Occasionally, due to how a module is constructed, jest.spyOn will not work.
```

Comment:
No sweat if you don't know and I could probably search around but do you know off the top of your head what about the module construction makes it so that we need this? I've been thinking about creating some kind of test cheatsheet for myself and to share out which could be a good place to note it.

Reply:
I actually don't quite get it and need to dig in more. Some of the error messages make it sounds like the implementations are marked as readonly. I prefer the spyOn approach, so I might file a follow up PR if I can chnage models to support spyOn.